### PR TITLE
Bluetooth: Mesh: Add Config Client API

### PR DESCRIPTION
**Pull request by haoyanbiao**
The api is used to reset a node (other than a Provisioner) and
remove it from the network

Signed-off-by: YanBiao Hao <haoyanbiao@xiaomi.com>

### CODE REVIEW

File: subsys/bluetooth/mesh/cfg_cli.c [hunk 1]
```diff
@@ -1125,6 +1140,39 @@ int bt_mesh_cfg_app_key_add(uint16_t net_idx, uint16_t addr, uint16_t key_net_id
...
+	BT_MESH_MODEL_BUF_DEFINE(msg, OP_NODE_RESET, 2);
+	struct bt_mesh_msg_ctx ctx = {
+		.net_idx = net_idx,
+		.app_idx = BT_MESH_KEY_DEV_REMOTE,
+		.addr = addr,
+		.send_ttl = BT_MESH_TTL_DEFAULT,
+	};
+	int err;
+
+	err = cli_prepare(NULL, OP_NODE_RESET_STATUS);
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/cfg_cli.c:1154]: Should get the status from the response status, here and in reset status callback.

File: doc/reference/bluetooth/mesh/shell.rst [hunk 2]
```diff
@@ -134,6 +134,12 @@ General configuration
 	Reset the mesh node to its initial unprovisioned state.
 
 
+``mesh node reset``
+-------------------
+
+        Reset a node (other than a Provisioner) and remove it from the network
```
Comment (trond-snekvik) [inline doc/reference/bluetooth/mesh/shell.rst:140]: Please indent with a single tab character

File: subsys/bluetooth/mesh/shell.c [hunk 3]
```diff
@@ -2607,6 +2607,22 @@ static int cmd_cdb_app_key_del(const struct shell *shell, size_t argc,
...
+	err = bt_mesh_cfg_node_reset(net.net_idx, net.dst, &reset);
+	if (err) {
+		shell_error(shell, "Unable to send Node Reset (err %d)", err);
+		return 0;
+	}
+
+	shell_print(shell, "Node reset complete");
+
+	return 0;
+}
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/shell.c:2624]: Could this be merged with the reset command instead? ie keep the current reset command, but do this inside. It should be possible to reset the local device with the config message, and having two very similar commands is confusing.

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

File: doc/reference/bluetooth/mesh/shell.rst [hunk 4]
```diff
@@ -128,10 +128,11 @@ General configuration
...
 
 
-``mesh reset``
---------------
+``mesh reset <value: local, remote>``
+-------------------------------------
 
-	Reset the mesh node to its initial unprovisioned state.
+	Reset the local mesh node to its initial unprovisioned state or reset a remote node and remove it from the network.
+	* ``value``: Sets local node or remote node.
```
Comment (trond-snekvik) [inline doc/reference/bluetooth/mesh/shell.rst:135]: The local/remote decision is made by the `dst` address in all other APIs, I think this should follow the same pattern. 
The command should also be moved into the Configuration client section.

File: subsys/bluetooth/mesh/cfg_cli.c [hunk 5]
```diff
@@ -1125,6 +1150,42 @@ int bt_mesh_cfg_app_key_add(uint16_t net_idx, uint16_t addr, uint16_t key_net_id
...
+		cli_reset();
+		return err;
+	}
+
+	if (!status) {
+		cli_reset();
+		return 0;
+	}
+
+	return cli_wait();
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/cfg_cli.c:1186]: As the status now only depends on whether the response message was received or not, it'll either be `true` or undefined, as we never explicitly set it to `false`. It should either be set to false before sending the message, or removed.

File: subsys/bluetooth/mesh/cfg_cli.c [hunk 6]
```diff
@@ -216,6 +216,30 @@ static void net_key_list(struct bt_mesh_model *model,
 	k_sem_give(&cli->op_sync);
 }
 
+struct node_reset_param {
+	bool *status;
+};
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/cfg_cli.c:221]: No need for this structure, really. The raw `bool *` can be used as an `op_param`.

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

**Comment by trond-snekvik** [comment]
In addition to the other comments: Please recheck your sign-off line format again. There should generally only be one, and it should match the commit author email. It should also be preceded by a blank line.

File: subsys/bluetooth/mesh/shell.c [hunk 7]
```diff
@@ -428,8 +428,31 @@ static int cmd_uuid(const struct shell *shell, size_t argc, char *argv[])
...
+			shell_print(shell, "invalid param :%s"
+					"local or remote is needed", argv[1]);
+			return 0;
+	}
+
+	if (!strcmp(argv[1], "local")) {
+		bt_mesh_reset();
+		shell_print(shell, "Local node reset complete");
+	} else {
+		uint8_t reset;
```
Comment (LingaoM) [inline subsys/bluetooth/mesh/shell.c:443]: Type bool here.

**Review by LingaoM — CHANGES_REQUESTED** [review]
Please check you sign-off line format, I’m notice that commit sign-off email inconsistent with git author.

Typically use `git commit -s` to input.

For more information, please see [here](https://docs.zephyrproject.org/latest/contribute/index.html#contribution-guidelines) in detail.

File: subsys/bluetooth/mesh/cfg_cli.c [hunk 8]
```diff
@@ -216,6 +216,26 @@ static void net_key_list(struct bt_mesh_model *model,
 	k_sem_give(&cli->op_sync);
 }
 
+static void node_reset_status(struct bt_mesh_model *model,
+		struct bt_mesh_msg_ctx *ctx, struct net_buf_simple *buf)
+{
+	struct bool *param = NULL;
```
Comment (LingaoM) [inline subsys/bluetooth/mesh/cfg_cli.c:222]: `struct  bool *param = NULL;` and below `*param->status`

**Review by LingaoM — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/cfg_cli.c (hunk 8 above)
Comment (LingaoM) [reply subsys/bluetooth/mesh/cfg_cli.c:222]: `bool` is a standard type defined by C programs, and other type declarations are not allowed as prefixes.

Comment (LingaoM) [reply subsys/bluetooth/mesh/cfg_cli.c:222]: You have some build error, please click Details in `buildkite/zephyr` below for more information.

**Review by trond-snekvik — CHANGES_REQUESTED** [review]
The status parameter still needs to be set to `false` inside `bt_mesh_cfg_node_reset`. Other than that, this LGTM.

**Comment by trond-snekvik** [comment]
@haoyanbiao if you don't mind, I'll amend the commit to clear the status, so we can get this merged

**Comment by haoyanbiao** [comment]
> @haoyanbiao if you don't mind, I'll amend the commit to clear the status, so we can get this merged

There is a lot of work to do recently. sorry about that I didn't check your information in time. Of course, you can modify it directly

**Review by trond-snekvik — APPROVED** [review]
I pushed a change to make sure status is always set now. The rest LGTM :+1:

File: subsys/bluetooth/mesh/shell.c [hunk 9]
```diff
@@ -428,8 +428,30 @@ static int cmd_uuid(const struct shell *shell, size_t argc, char *argv[])
...
+		shell_print(shell, "Local node reset complete");
+	} else {
+		int err;
+		bool reset = false;
+
+		err = bt_mesh_cfg_node_reset(net.net_idx, net.dst, &reset);
+		if (err) {
+			shell_error(shell, "Unable to send"
+					"Remote Node Reset (err %d)", err);
+			return 0;
```
Comment (LingaoM) [inline subsys/bluetooth/mesh/shell.c:449]: Is a space missing in the `shell_error` ?

File: subsys/bluetooth/mesh/cfg_cli.c [hunk 10]
```diff
@@ -1125,6 +1146,44 @@ int bt_mesh_cfg_app_key_add(uint16_t net_idx, uint16_t addr, uint16_t key_net_id
 	return cli_wait();
 }
 
+int bt_mesh_cfg_node_reset(uint16_t net_idx, uint16_t addr, bool *status)
+{
+	BT_MESH_MODEL_BUF_DEFINE(msg, OP_NODE_RESET, 2);
```
Comment (LingaoM) [inline subsys/bluetooth/mesh/cfg_cli.c:1151]: Config Node Reset no need for payloads, of course, the length.
`BT_MESH_MODEL_BUF_DEFINE(msg, OP_NODE_RESET, 0);`

**Comment by LingaoM** [comment]
Other LGTM 👍

**Review by LingaoM — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**